lib/Dialect/Synth/Transforms/CutRewriter.cpp

```diff
@@ -0,0 +1,564 @@
+
+bool Cut::isTrivialCut() const {
+  // A cut is a primary input if it has no operations and only one input
+  return operations.empty() && inputs.size() == 1;
+}
+
+mlir::Operation *Cut::getRoot() const {
+  return operations.empty()
+             ? nullptr
+             : operations.back(); // The last operation is the root
+}
+
+const NPNClass &Cut::getNPNClass() const {
+  // If the NPN is already computed, return it
+  if (npnClass)
+    return *npnClass;
+
+  auto truthTable = getTruthTable();
+
+  // Compute the NPN canonical form
+  auto canonicalForm = NPNClass::computeNPNCanonicalForm(truthTable);
+
+  npnClass.emplace(std::move(canonicalForm));
+  return *npnClass;
+}
+
+void Cut::getPermutatedInputs(const NPNClass &patternNPN,
+                              SmallVectorImpl<Value> &permuatedInputs) const {
+  auto npnClass = getNPNClass();
+  SmallVector<Value> permutedInputs;
+  SmallVector<unsigned> idx;
+  npnClass.getInputPermutation(patternNPN, idx);
+  permuatedInputs.reserve(idx.size());
+  for (auto inputIndex : idx) {
+    assert(inputIndex < inputs.size() && "Input index out of bounds");
+    permuatedInputs.push_back(inputs[inputIndex]);
+  }
+}
+
+void Cut::dump(llvm::raw_ostream &os) const {
+  os << "// === Cut Dump ===\n";
+  os << "Cut with " << getInputSize() << " inputs and " << operations.size()
+     << " operations:\n";
+  if (isTrivialCut()) {
+    os << "Primary input cut: " << *inputs.begin() << "\n";
+    return;
+  }
+
+  os << "Inputs: \n";
+  for (auto [idx, input] : llvm::enumerate(inputs)) {
+    os << "  Input " << idx << ": " << input << "\n";
+  }
+  os << "\nOperations: \n";
+  for (auto *op : operations) {
+    op->print(os);
+    os << "\n";
+  }
+  auto &npnClass = getNPNClass();
+  npnClass.dump(os);
+
+  os << "// === Cut End ===\n";
+}
+
+unsigned Cut::getInputSize() const { return inputs.size(); }
+
+unsigned Cut::getOutputSize() const { return getRoot()->getNumResults(); }
+
+const BinaryTruthTable &Cut::getTruthTable() const {
+  if (truthTable)
+    return *truthTable;
+
+  if (isTrivialCut()) {
+    // For a trivial cut, a truth table is simply the identity function.
+    // 0 -> 0, 1 -> 1
+    truthTable = BinaryTruthTable(1, 1, {llvm::APInt(2, 2)});
+    return *truthTable;
+  }
+
+  int64_t numInputs = getInputSize();
+  int64_t numOutputs = getOutputSize();
+  assert(numInputs < 20 && "Truth table is too large");
+
+  // Simulate the IR.
+  uint32_t tableSize = 1 << numInputs;
+  DenseMap<Value, APInt> eval;
+  for (uint32_t i = 0; i < numInputs; ++i) {
+    APInt value(tableSize, 0);
+    for (uint32_t j = 0; j < tableSize; ++j) {
+      // Make sure the order of the bits is correct.
+      value.setBitVal(j, (j >> i) & 1);
+    }
+    // Set the input value for the truth table
+    eval[inputs[i]] = std::move(value);
+  }
+
+  // Simulate the operations in the cut
+  for (auto *op : operations)
+    simulateLogicOp(op, eval);
```

Comment:
Very cool idea to generate these input truth table vectors and then flow them through the ops, essentially vectorizing the circuit evaluation using bitwise logic ops. 🥳